dbcon/joblist/diskjoinstep.cpp

```diff
@@ -260,30 +249,85 @@ void DiskJoinStep::largeReader()
     abort();
   }
 
-  if (cancelled())
+  if (UNLIKELY(cancelled()))
     while (more)
       more = largeDL->next(largeIt, &rgData);
 }
 
-void DiskJoinStep::loadFcn()
+void DiskJoinStep::loadFcn(const uint32_t threadID, const uint32_t smallSideSizeLimit,
+                           const std::vector<joiner::JoinPartition*>& joinPartitions)
 {
   boost::shared_ptr<LoaderOutput> out;
-  bool ret;
 
   try
   {
-    do
+    uint32_t partitionIndex = 0;
+    bool partitionDone = true;
+
+    // Iterate over partitions.
+    while (partitionIndex < joinPartitions.size() && !cancelled())
     {
+      uint64_t currentSize = 0;
+      auto* joinPartition = joinPartitions[partitionIndex];
       out.reset(new LoaderOutput());
-      ret = jp->getNextPartition(&out->smallData, &out->partitionID, &out->jp);
 
-      if (ret)
+      if (partitionDone)
+        joinPartition->setNextSmallOffset(0);
+
+      while (true)
       {
-        // cout << "loaded partition " << out->partitionID << " smallData = " << out->smallData.size() <<
-        // endl;
-        loadFIFO->insert(out);
+        messageqcpp::ByteStream bs;
+        RowGroup rowGroup;
+        RGData rgData;
+
+        joinPartition->readByteStream(0, &bs);
+        if (!bs.length())
+        {
+          partitionDone = true;
+          break;
+        }
+
+        rgData.deserialize(bs);
+        rowGroup.setData(&rgData);
+
+        // Check that current `RowGroup` has rows.
+        if (!rowGroup.getRowCount())
+        {
+          partitionDone = true;
+          break;
+        }
+
+        currentSize += rowGroup.getRowCount() * rowGroup.getColumnCount() * 16;
```

Comment:
This heuristic isn't clear for me. Why do you decide to go with 16  for a column width? There are methods to retrieve the actual RGData size.

Reply:
Thanks, fixed this, looks like a bug.